### PR TITLE
V4 tracker

### PR DESCRIPTION
**Pull request by s1na**
This PR tracks the changes planned for v4 (see #455). It also acts as a base branch for the pull requests, to prevent mixing experimental development in `master`. This'd allow us to maintain v3, fix bugs and do releases independently of the changes made here, and also have parallel release candidates for v4.

- [x] Add promisified wrapper for stateManager (#480)
- [x] Class-based promisified rewrite of runCall and runCode (#483)
- [x] Use EEI methods in opFns (#486)
- [x] Promisify async opcode handlers (#491)
- [x] Detach precompiles from VM (#492)
- [x] Promisify runTx internals (#493)
- [x] Rewrite runBlock internals with async/await (#494)
- [x] Setup gradual process for migrating to Typescript (#495)
- [x] Migrate state module to TS (#496)
- [x] Migrate precompiles, opcodes, eei, message, txContext to typescript (#497)
- [x] Subdivide runState, refactor loop (#498)
- [x] Split executeMessage into executeCall, executeCreate (#499)
- [x] Migrate interpreter & exceptions to typescript (#504)
- [x] Migrate Loop to typescript (#505)
- [x] Migrate opFns to typescript, modify loop/interpreter structure (#506)
- [x] Migrate index to typescript (#507)
- [x] Migrate runCode to typescript (#508)
- [x] Migrate runCall to typescript (#510)
- [x] Migrate runTx to typescript (#511)
- [x] Migrate runBlock to typescript (#512)
- [x] Drop runJit (#515)
- [x] Migrate runBlockchain to typescript (#517)
- [x] Use complete typescript configs, fix linting/formatting errors (#518)

### CODE REVIEW

File: lib/evm/interpreter.js [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+      code: message.code,
+      data: message.data,
+      gasLimit: message.gasLimit,
+      address: message.to,
+      caller: message.caller,
+      value: message.value,
+      depth: message.depth,
+      selfdestruct: message.selfdestruct,
+      static: message.isStatic
+    }
```
Comment (holgerd77) [inline lib/evm/interpreter.js]: What is the reason for this split of txContext and message? Couldn't this be combined in the first place, since this is combined here again?

Comment (s1na) [reply lib/evm/interpreter.js]: Splitting this `opts` into `message` and `txContext` (+ `storageReader` and `block` for now) is something I'm planning to do along with splitting `runState`.

File: lib/evm/interpreter.js [hunk 2]
```diff
@@ -0,0 +1,206 @@
+const promisify = require('util.promisify')
+const BN = require('bn.js')
+const ethUtil = require('ethereumjs-util')
+const { ERROR } = require('../exceptions')
+const Loop = require('./loop')
+const { StorageReader } = require('../state')
+const PStateManager = require('../state/promisified')
+
+module.exports = class Interpreter {
```
Comment (holgerd77) [inline lib/evm/interpreter.js]: Would it make sense to make the Interpreter class just the index.js entry point of the whole module and therefore ease the API? Or is there a reason to keep this explicit?

Comment (s1na) [reply lib/evm/interpreter.js]: It's definitely possible if you'd prefer. I'm following the convention that `index` contains (almost) no source code, and only is there to expose code to other modules. Other modules would then only import what's exported in `index` and not individual source files.

File: lib/evm/interpreter.js [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+  constructor (vm, txContext, block, storageReader) {
```
Comment (holgerd77) [inline lib/evm/interpreter.js]: General note: I think every access through _vm here is an indicator that we have some misplacement of code or suboptimal structure.

File: lib/evm/interpreter.js (hunk 1 above)
Comment (holgerd77) [reply lib/evm/interpreter.js]: Ah ok.

File: lib/evm/interpreter.js (hunk 2 above)
Comment (holgerd77) [reply lib/evm/interpreter.js]: Ok, let's keep for now. Still thinking that at the end people will instantiate their own (eventually modified/extended) EVM or eWASMVM, which they will pass to the outer VM execution engine similar to the StateManager being passed now, and where they will likely also directly get the events from (so we need no binding to the outer VM instance any more). Is this a similar model you envisioned, or do you have got an alternative structure in mind?

Comment (s1na) [reply lib/evm/interpreter.js]: Yeah that's quite similar to what I was thinking. VM would bind all these different classes together (e.g. evm/ewasm interpreter, state manager, etc.). Users could instantiate VM possibly with their custom implementation of these classes, and call higher-level functions like `runTx` or `runBlock`, or they could use the classes themselves if they need only a specific aspect of the logic.

**Comment by holgerd77** [comment]
We should eventually start to develop a rough idea when we might be ready to release here. No rush but it would at some point be good to have this merged to ``master`` to have a solid basis for ``Istanbul`` development, think it would make no sense to develop EIP integrations towards the old ``master`` and then try to re-integrate.

@s1na Do you think it might be realistic to have some consistent state in let's say 3-4 weeks we could do a release upon? What are the things you want to do further do work upon respectively you very much want to have integrated before?

We can also think about the release process itself, and e.g. start with releasing ``alpha`` versions or the like if it makes sense.

**Comment by s1na** [comment]
Very much agree. I think 3rd weekend of May should be a good target (also coincides with Istanbul proposals being accepted). I'll try to take a broader view and see what could be done in these three weeks.

I'd however appreciate some help on the external API side, i.e. if we want to make any changes there and if so what changes (e.g. do we want to promisify the whole API?). Also pinging @alcuadrado for feedback on this.

For release I think first doing a beta release and only after merging to master makes sense.

**Comment by holgerd77** [comment]
Sounds great, thanks! 😄

The API if the VM is actually pretty simple - just 4-5 methods like ``runCall``and the like and promises are really not brand new technologies any more.

I am turning more and more to not doing endless deprecation rounds but throwing the old stuff away and doing a clean new promisifed API. Shouldn't be too much of a burden for people to upgrade.

**Comment by holgerd77** [comment]
@s1na Will put some additional thought on this during the next days, but @alcuadrado and others, your input and concrete suggestions are also super welcome on this!

**Comment by s1na** [comment]
Regarding external API another point is what do we want to expose (from the new classes for example), and how do we facilitate users customizing behavior. To give an example, to address the use-cases stated in #410 by @pinkiebell, `Interpreter` and `Loop` should also be exposed and accepted as a parameter in either `VM` or `runTx`.

**Comment by holgerd77** [comment]
Can we come up with some consistent definitions what an ``Interpreter`` and a ``Loop`` actually are and what the scope of responsibility is for each class (and also add theses to the class TSDoc comments)?

I think this would help a lot with the API definition and demarcation.

**Comment by holgerd77** [comment]
Another approach: we don't expose very much new API at all in ``v4`` - @s1na since you kept all this 
farsightedly very much backwards-compatible - and we just expose people (more) to the new code and internal structures written (exposing by just: releasing the respective code) - and let this settle more over time and further evolve, than eventually also with more input from the broader community. And then do the API in ``v5``.

Maybe not as satifsfying as a more radical approach but eventually more practical and long-term sustainable.

**Comment by holgerd77** [comment]
But if we come up with a solution consistent, simple and thought through enough we can also do it in ``v4``. Let's see. 😄

**Comment by holgerd77** [comment]
I am generally not sure about the ``loop`` naming, would ``handler`` or ``processor`` or similar be more fitting? Is this really a ``loop``?

**Comment by holgerd77** [comment]
I think definitions / task descriptions are also needed for ``EEI`` and ``Message``.

**Comment by holgerd77** [comment]
@s1na Is this class-based opcode structure inheriting from a generic ``Opcode`` class implemented in ``evm-ts`` (see [opcodes](https://github.com/ethereum-ts/evm-ts/tree/master/lib/opcodes) folder) something we want to take inspiration from? Then people can subclass specific opcodes and do their own modifications. Or is this too problematic / comes with too much uncertainties on the performance side?

//cc @krzkaczor

**Comment by krzkaczor** [comment]
Thanks for calling me out @holgerd77 😆 

I think that having opcodes implemented as classes was a design mistake in `evm-ts`. What should be done instead is having huge tagged union describing all possible opcodes and then an `execute(state: State, opcode: Opcode): State` function that just executes it and returns a new state. TypeScript compiler can warn you if there are some unhandled opcodes.

This way you can easily add new behavior to existing opcodes — it's just a function taking opcode as an argument and doing whatever you want with it for example pretty print bytecode.

Changing `execute` behavior (this user plugins) can be implemented with middleware like concept.

In summary, it's pretty similar to `opFns` but typesafe.

**Comment by pinkiebell** [comment]
I think having an Interpreter class  which can be either subclassed or passed in some delegate functions is fine. Though we have to take care about `call depth` too.

class Interpreter
  - has all the opcode names as function prototypes >  can be overriden
  - has a basic runLoop > can be overriden
  - inside the runLoop it may do some gardening/guarding stuff before dispatching to the opcode hander and it might be nice to have an intermediate overridable function like `handleNextInstruction`

**Comment by holgerd77** [comment]
@s1na Just had a broader look through the different classes in ``evm``. I have the impression that there are there are still many open questions on the concrete shaping of message signatures respectively interdependencies between components, responsibilities for components respectively the location where things should be handled, so things on the internal structure.

Then there are the questions on what should live within ``evm`` and what might be generic enough to drag out (``EEI`` (with some losses)? Memory?

And then we have the event submission structure questions like what events should be there, who should submit, stuff like that.

For me that sounds too much to figure out until the next release. I have some tendency therefore to do that release with the API completely unchanged and get out the code into the wild. Once we work towards ``v5`` more people will get involved with the code and there might be some natural discussion about structure popping up and we can give this more time to evolve.

Then we could concentrate within this round on finishing the ``TypeScript`` changes, cleanup and tying things together.

Not sure, does this make sense? We can also leave this open a bit longer and maybe wait what the next PRs bring. I have to admit that I can't really judge what short-term changes you still have in mind and what is still only temporarily added.

**Comment by holgerd77** [comment]
We can complement this with two additional options ``opts.interpreter`` and ``opts.loop`` and mark as ``EXPERIMENTAL`` so that people have better possibility to experiment on this. Not sure if this is responsible though if things are still under too heavy change.

**Comment by holgerd77** [comment]
Will list my main points on interdependencies and locations - without having another look at what instances are used for within the classes - but just out of my "structure intuition" 😋:

- ``EEI`` should be passed as an option to ``Interpreter`` from the outside
- ``Interpreter`` shouldn't have a dependency to ``_vm``, likely a ``Common`` instance should be passed instead (to cover some of the usages, others need other solutions)
- ``Loop`` should have no dependency to both ``VM`` and ``Interpreter``
- ``Interpreter`` (and/or ``Loop``) should directly emit the events

Not sure, definitely don't take this as a TODO list 😄. Some of these changes might not be justified or make sense at all, but others might lead to further refactoring/restructuring changes and point to currently suboptimal/misplaced structural components which prevent these changes in the first place.

**Comment by s1na** [comment]
> I have some tendency therefore to do that release with the API completely unchanged and get out the code into the wild. Once we work towards v5

I think that makes sense. I will just try to tidy up, focus mostly on typescript transition and fixing some of the structural dependencies you mentioned. Other than that I don't think we can make larger changes in time. But hopefully we can try to plan more regular releases like this.

**Comment by holgerd77** [comment]
@s1na Ok, if you also have such a clear opinion on this then let's move forward like you described. I've no problems with doing more regular major releases if we have changes/updates which justify them respectively make them necessary.

**Comment by startupgurukul** [comment]
Looks like v4 release is on the horizon. Looking forward to it.

Visiting the repo after long time & you have made awesome strides with ethereum VM in JS, making ethereum development very smooth and hopefully helping blockchain reach its true potential

**Comment by holgerd77** [comment]
@startupgurukul thanks! 😀

**Comment by holgerd77** [comment]
@s1na if you feel comfortable with that I would suggest we try to do the ``master`` merge here on Monday and do subsequent cleanup work towards the ``master`` branch. Then we have two days where the dust can settle on the ``master`` branch and we see through a bit again on Wednesday. 😄 😋

Can you do some ``master`` rebase here (this would be the procedure, right?)?

Let me know if you have any reservations, in doubt we are not completely bound to this date in any way and should rather make sure that the process is robust and we are confident with the integration work.

**Comment by s1na** [comment]
That sounds good to me. Only would it be okay if I merge master into this branch, and later on we merge v4 into master (instead of rebase)? Due to the large number of commits, and because there are conflicts with master, rebasing can get difficult. I also took care to squash and merge the PRs, so we don't have tons of minor commits polluting the commit history.

**Comment by holgerd77** [comment]
Ok, that sounds good!

**Comment by s1na** [comment]
I have merged master into `v4`. Should be ready for merging :smile:

**Comment by startupgurukul** [comment]
Yay can you also push the latest to npm pls

**Comment by holgerd77** [comment]
@startupgurukul probably two days if nothing unexpected trigger a delay until a beta release on npm

**Comment by holgerd77** [comment]
@s1na `npm run testStatePetersburg` is actually failing for me on `master` for all tests, even with a clean `npm install` + previously delete `node_modules` folder. Some example output:

```shell
$ npm run testStatePetersburg

> ethereumjs-vm@3.0.0 testStatePetersburg
> npm run build:dist && node ./tests/tester -s --fork='Petersburg' --dist

> ethereumjs-config-build

+ exec tsc -p ./tsconfig.prod.json
TAP version 13
# GeneralStateTests
# file: addNonConst test: addNonConst
not ok 1 the state roots should match
  ---
    operator: equal
    expected: |-
      '2d6da043f5cfd9fe3b1feab48970d855bf5570f5316492869642b228d8bed3ca'
    actual: |-
      '3c3f05727bc49ff870f5e780c5d00426d9c153de7e23443e70ec2da43d7f731c'
    at: replenish (/Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/async/dist/async.js:998:17)
  ...
not ok 2 the state roots should match
  ---
    operator: equal
    expected: |-
      'e396abd1e58496cc86a1a5bc31af70c98926cceef3738e1189af4c226b61da15'
    actual: |-
      '3c3f05727bc49ff870f5e780c5d00426d9c153de7e23443e70ec2da43d7f731c'
    at: replenish (/Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/async/dist/async.js:998:17)
  ...
```

Any idea?

**Comment by s1na** [comment]
@holgerd77 Hmm :thinking: I tried on a fresh clone and the tests seem to pass. I wonder what could be the cause. Could you maybe add an `console.log(err)` here:

https://github.com/ethereumjs/ethereumjs-vm/blob/9f96231cf7b4b185368092416aa7af2c30d2f90b/lib/runTx.ts#L95-L96

to see if any error is occuring?

**Comment by holgerd77** [comment]
@s1na Ah yes, great hint:

```shell
+ exec tsc -p ./tsconfig.prod.json
TAP version 13
# GeneralStateTests
# file: addNonConst test: addNonConst
TypeError: bloom_1.default is not a constructor
    at txLogsBloom (/ethereumjs-vm/dist/runTx.js:250:17)
    at VM.<anonymous> (/ethereumjs-vm/dist/runTx.js:157:37)
    at step (/ethereumjs-vm/dist/runTx.js:32:23)
    at Object.next (/ethereumjs-vm/dist/runTx.js:13:53)
    at fulfilled (/ethereumjs-vm/dist/runTx.js:4:58)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:189:7)
not ok 1 the state roots should match
  ---
    operator: equal
    expected: |-
      '2d6da043f5cfd9fe3b1feab48970d855bf5570f5316492869642b228d8bed3ca'
    actual: |-
      '3c3f05727bc49ff870f5e780c5d00426d9c153de7e23443e70ec2da43d7f731c'
    at: replenish (/node_modules/async/dist/async.js:998:17)
```

**Comment by s1na** [comment]
Maybe the typescript compiler hasn't overwritten previously built files. Could you try removing `dist` and then running the tests?

**Comment by holgerd77** [comment]
Ah, yes, works: 😄
